### DISCUSSION TITLE
Occasional exceptions "Trying to access array offset on value of type null"

### DISCUSSION THREAD

**Discussion by illambo** [stackkit/laravel-google-cloud-tasks-queue#29] [Q&A] [answered]
Hello, first of all thanks for the package, much appreciated! 

I'm checking for these sporadic exceptions _"Trying to access array offset on value of type null"_

- App running on Google Cloud Run 
- PHP 7.4.14 
- laravel/framework 8.31.0
- stackkit/laravel-google-cloud-tasks-queue 2.0.1
- grpc disabled in order to prevent many _"Google\ApiCore\ApiException: Socket closed"_ errors (see https://github.com/googleapis/google-cloud-php/issues/2427)

Example of two common related stack traces exceptions:

> ErrorException: Trying to access array offset on value of type null
> #39 /vendor/stackkit/laravel-google-cloud-tasks-queue/src/OpenIdVerificator.php(74): Illuminate\Foundation\Bootstrap\HandleExceptions::handleError
> #38 /vendor/stackkit/laravel-google-cloud-tasks-queue/src/OpenIdVerificator.php(74): Stackkit\LaravelGoogleCloudTasksQueue\OpenIdVerificator::extractPublicKeyFromCertificate
> #37 /vendor/stackkit/laravel-google-cloud-tasks-queue/src/OpenIdVerificator.php(62): Stackkit\LaravelGoogleCloudTasksQueue\OpenIdVerificator::getPublicKey
> #36 /vendor/stackkit/laravel-google-cloud-tasks-queue/src/OpenIdVerificator.php(38): Stackkit\LaravelGoogleCloudTasksQueue\OpenIdVerificator::decodeOpenIdToken
> #35 /vendor/stackkit/laravel-google-cloud-tasks-queue/src/TaskHandler.php(51): Stackkit\LaravelGoogleCloudTasksQueue\TaskHandler::authorizeRequest
> #34 /vendor/stackkit/laravel-google-cloud-tasks-queue/src/TaskHandler.php(32): Stackkit\LaravelGoogleCloudTasksQueue\TaskHandler::handle
> #33 /vendor/laravel/framework/src/Illuminate/Routing/ControllerDispatcher.php(48): Illuminate\Routing\ControllerDispatcher::dispatch
> [...]

> ErrorException: Trying to access array offset on value of type null
> #56 /vendor/google/gax/src/Transport/RestTransport.php(161): Illuminate\Foundation\Bootstrap\HandleExceptions::handleError
> #55 /vendor/google/gax/src/Transport/RestTransport.php(161): Google\ApiCore\Transport\RestTransport::convertToApiException
> #54 /vendor/google/gax/src/Transport/RestTransport.php(131): Google\ApiCore\Transport\RestTransport::Google\ApiCore\Transport\{closure}
> #53 /vendor/guzzlehttp/promises/src/Promise.php(204): GuzzleHttp\Promise\Promise::callHandler
> #52 /vendor/guzzlehttp/promises/src/Promise.php(153): GuzzleHttp\Promise\Promise::GuzzleHttp\Promise\{closure}
> #51 /vendor/guzzlehttp/promises/src/TaskQueue.php(48): GuzzleHttp\Promise\TaskQueue::run
> #50 /vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php(158): GuzzleHttp\Handler\CurlMultiHandler::tick
> #49 /vendor/guzzlehttp/guzzle/src/Handler/CurlMultiHandler.php(183): GuzzleHttp\Handler\CurlMultiHandler::execute
> #48 /vendor/guzzlehttp/promises/src/Promise.php(248): GuzzleHttp\Promise\Promise::invokeWaitFn
> #47 /vendor/guzzlehttp/promises/src/Promise.php(224): GuzzleHttp\Promise\Promise::waitIfPending
> #46 /vendor/guzzlehttp/promises/src/Promise.php(269): GuzzleHttp\Promise\Promise::invokeWaitList
> #45 /vendor/guzzlehttp/promises/src/Promise.php(226): GuzzleHttp\Promise\Promise::waitIfPending
> #44 /vendor/guzzlehttp/promises/src/Promise.php(62): GuzzleHttp\Promise\Promise::wait
> #43 /vendor/google/cloud/Tasks/src/V2/Gapic/CloudTasksGapicClient.php(1374): Google\Cloud\Tasks\V2\Gapic\CloudTasksGapicClient::createTask
> #42 /vendor/stackkit/laravel-google-cloud-tasks-queue/src/CloudTasksQueue.php(74): Stackkit\LaravelGoogleCloudTasksQueue\CloudTasksQueue::pushToCloudTasks
> #41 /vendor/stackkit/laravel-google-cloud-tasks-queue/src/CloudTasksQueue.php(36): Stackkit\LaravelGoogleCloudTasksQueue\CloudTasksQueue::push
> #40 /vendor/laravel/framework/src/Illuminate/Bus/Dispatcher.php(251): Illuminate\Bus\Dispatcher::pushCommandToQueue
> [...]

I think it mostly depended (seems clear) on the underline communication with google but is there a way to better handle it?
Do you have any suggestions?
Meanwhile, I try to raise the version of the library and manage the thing with the failed job.

Thanks!

**Comment by marickvantuil** [comment]
> Hello, first of all thanks for the package, much appreciated!

Thank you!


If I understood correctly, some of Google's endpoints that this package uses throw random 'Socket closed' errors? That's strange, I haven't seen that happening yet, though I haven't used this package extensively yet (hopefully soon).

It seems to happen when verifying the OpenID token, as well as when sending the queued job to Cloud Tasks? Does it work if you try it right after? Maybe we could build some retry mechanism for this issue?

**Reply by illambo** [reply]
Thanks for the reply,
with grpc enabled the "Occasional exceptions" were of the typology _"Socket closed"_ (see my comment [here](https://github.com/googleapis/google-cloud-php/issues/2427#issuecomment-776592676)),

example below, I used a simple webhook controller for dispatch my "cron" jobs with Google Cloud Scheduler 

> Google\ApiCore\ApiException: {
>     "message": "Socket closed",
>     "code": 14,
>     "status": "UNAVAILABLE",
>     "details": []
> }
>   File "/app/Http/Controllers/Webhooks/ScheduleController.php", line 48, in App\Http\Controllers\Webhooks\ScheduleController::processPendingPayables
>     ProcessPendingPayables::dispatch();
>   File "/public/index.php", line 52
>     $request = Request::capture()
> ...
> (43 additional frame(s) were not displayed)

So I tried disabling grpc and the exceptions "converted" to the ones given above in the discussion _("Trying to access array offset on value of type null")_

Now I monitor for a few days how an environment that I have in production behaves with the updated version of the library and in case I think about some retry / try catch loop as you suggested.

In the meantime, thank you.

It remains incomprehensible to me how google can have such issues like googleapis/google-cloud-php#2427 and googleapis/google-cloud-php#2338 open for years...it is said that time heals all the wounds... :laughing:

**Comment by illambo** [comment]
Ok I think i found something related to [OpenIdVerificator getPublicKey method](https://github.com/stackkit/laravel-google-cloud-tasks-queue/blob/453d9a75d149a04e2a749a0de6f9b0e5dbe4e7d6/src/OpenIdVerificator.php#L51),
following my stack trace errors and executed queries I can see in order (related to GOOGLE_V3_CERTS):

- select * from `cache` where `key` = ? limit 1
- select * from `cache` where `key` = ? limit 1
- delete from `cache` where `key` = ?

I think it maybe an expiration problem, the condition `Cache::has(self::V3_CERTS)` return true but the next get `$v3Certs = Cache::get(self::V3_CERTS);` it can return null.

This would explain the error and the reason why _getFreshCertificates_ method is not run.

What do you think?

**Reply by illambo** [reply]
I opened a #30 PR if you find it useful.

**Comment by marickvantuil** [comment, marked as answer]
Thanks, definitely sounds like it could cause problems. I merged the PR and tagged `v2.1.3`. Can you check if this fixes the issues?

**Reply by illambo** [reply]
Certainly, I was already testing it and for now I no longer encounter this problem. 
I'll update you if it occurs. 
Thank you!

**Reply by illambo** [reply]
Hi, I have been monitoring in these two / three weeks and the problem (which previously occurred with an incidence of about once a day) has never occurred to me anymore. So great, thank you.
